humility-core/src/arch.rs

```diff
@@ -2,339 +2,176 @@
+                    rval.push(Register::RiscV(reg));
+                }
+            }
+            _ => unimplemented!(),
         }
     }
 
     rval
 }
 
-fn instr_source_target(
+pub fn instr_source_target(
     cs: &Capstone,
     instr: &capstone::Insn,
-) -> Result<(Option<ARMRegister>, Option<ARMRegister>)> {
+) -> Result<(Option<Register>, Option<Register>)> {
     let detail = cs.insn_detail(instr).unwrap();
-
-    let mut source: Option<ARMRegister> = None;
-    let mut target: Option<ARMRegister> = None;
+    let mut source_id: Option<u16> = None;
+    let mut source: Option<Register> = None;
+    let mut target_id: Option<u16> = None;
+    let mut target: Option<Register> = None;
 
     for op in detail.regs_read() {
-        if source.is_some() {
+        if source_id.is_some() {
             bail!("multiple source registers");
         }
-        source = Some((*op).into());
+        let RegId(op) = op;
+        source_id = Some(*op);
     }
 
     for op in detail.regs_write() {
-        if target.is_some() {
+        if target_id.is_some() {
             bail!("multiple target registers");
         }
-        target = Some((*op).into());
+        let RegId(op) = op;
+        target_id = Some(*op);
     }
 
-    Ok((source, target))
-}
-
-//
-// On ARM, our stub frames (that is, those frames that contain system call
-// instructions) have no DWARF information that describes how to unwind
-// through them; for these frames we do some (very crude) analysis of the
-// program text to determine what registers are pushed and how they are
-// manipulated so we can properly determine register state before the system
-// call.  Note that this made slightly more challenging by ARMv6-M, which
-// doesn't have as rich push instructions as ARMv7-M/ARMv8-M:  in order for it
-// to push R8 through R11, it must first move them into the lower registers
-// (R0 through R7).  We therefore have to track moves in addition to pushes to
-// determine what landed where -- and yes, this heuristic is incomplete!
-//
-pub fn presyscall_pushes(
-    cs: &Capstone,
-    instrs: &[capstone::Insn],
-) -> Result<Vec<ARMRegister>> {
-    const ARM_INSN_PUSH: u32 = arch::arm::ArmInsn::ARM_INS_PUSH as u32;
-    const ARM_INSN_MOV: u32 = arch::arm::ArmInsn::ARM_INS_MOV as u32;
-    const ARM_INSN_POP: u32 = arch::arm::ArmInsn::ARM_INS_POP as u32;
-
-    let mut map = HashMap::new();
-    let mut rval = vec![];
-
-    for instr in instrs {
-        match instr.id() {
-            InsnId(ARM_INSN_MOV) => {
-                let (source, target) = instr_source_target(cs, instr)?;
-
-                if let (Some(source), Some(target)) = (source, target) {
-                    map.insert(target, source);
-                }
+    match detail.arch_detail() {
+        ArchDetail::ArmDetail(_detail) => {
+            if let Some(id) = source_id {
+                source =
+                    Some(Register::Arm(ARMRegister::from_u16(id).unwrap()));
```

Comment:
I have also seen this pattern in Hubris, although I'm not a very big fan. Personally, I think it's best if we avoid it, as it can just confuse things further.

Reply:
I just went ahead and rewrote it.  It is much cleaner now.